extension/maxminddb.c

```diff
@@ -389,6 +408,200 @@ static void Reader_dealloc(PyObject *self) {
+                left->depth = cur->depth + 1;
+                left->record = node.left_record;
+                left->type = node.left_record_type;
+                left->entry = node.left_record_entry;
+
+                struct record *right = left->next = calloc(1, sizeof(record));
+                if (right == NULL) {
+                    free(cur);
+                    free(left);
+                    PyErr_NoMemory();
+                    return NULL;
+                }
+                memcpy(
+                    right->ip_packed, cur->ip_packed, sizeof(right->ip_packed));
+                right->ip_packed[cur->depth / 8] |= 1 << (7 - cur->depth % 8);
+                right->depth = cur->depth + 1;
+                right->record = node.right_record;
+                right->type = node.right_record_type;
+                right->entry = node.right_record_entry;
+                right->next = ri->next;
+
+                ri->next = left;
+                break;
+            }
+            case MMDB_RECORD_TYPE_EMPTY:
+                break;
+            case MMDB_RECORD_TYPE_DATA: {
+                MMDB_entry_data_list_s *entry_data_list = NULL;
+                int status =
+                    MMDB_get_entry_data_list(&cur->entry, &entry_data_list);
+                if (MMDB_SUCCESS != status) {
+                    PyErr_Format(
+                        MaxMindDB_error,
+                        "Error looking up data while iterating over tree: %s",
+                        MMDB_strerror(status));
+                    MMDB_free_entry_data_list(entry_data_list);
+                    free(cur);
+                    return NULL;
+                }
+
+                MMDB_entry_data_list_s *original_entry_data_list =
+                    entry_data_list;
+                PyObject *record = from_entry_data_list(&entry_data_list);
+                MMDB_free_entry_data_list(original_entry_data_list);
+                if (record == NULL) {
+                    free(cur);
+                    return NULL;
+                }
+
+                int ip_start = 0;
+                int ip_length = 4;
+                if (ri->reader->mmdb->depth == 128) {
+                    if (is_ipv6(cur->ip_packed)) {
+                        // IPv6 address
+                        ip_length = 16;
+                    } else {
+                        // IPv4 address in IPv6 tree
+                        ip_start = 12;
+                    }
+                }
+                PyObject *network_tuple =
+                    Py_BuildValue("(y#i)",
+                                  &(cur->ip_packed[ip_start]),
+                                  ip_length,
+                                  cur->depth - ip_start * 8);
+                if (network_tuple == NULL) {
+                    Py_DECREF(record);
+                    free(cur);
+                    return NULL;
+                }
+                PyObject *args = PyTuple_Pack(1, network_tuple);
```

Comment:
Hmm, isn't it already a tuple we could use from the prior call?

Reply:
It is a tuple containing a tuple! :turtle: The args are a tuple and we are passing it a tuple containing the packed address plus the prefix length.